apps/api/src/app/environments-v1/usecases/output-renderers/expand-email-editor-schema.usecase.ts

```diff
@@ -28,60 +33,98 @@ export class ExpandEmailEditorSchemaUsecase {
+    const processedNode = structuredClone(rootNode);
+    await this.traverseAndProcessNodes(processedNode, variables);
+
+    return processedNode;
+  }
+
+  private async traverseAndProcessNodes(
     node: TipTapNode,
-    parentNode?: TipTapNode
-  ): Promise<TipTapNode> {
-    if (node.content) {
-      const processedContent: TipTapNode[] = [];
-      for (const innerNode of node.content) {
-        const processed = await this.processShowAndForControls(variables, innerNode, parentNode);
-        if (processed) {
-          processedContent.push(processed);
+    variables: Record<string, unknown>,
+    parent?: TipTapNode
+  ): Promise<void> {
+    const queue: Array<{ node: TipTapNode; parent?: TipTapNode }> = [{ node, parent }];
+
+    while (queue.length > 0) {
+      const current = queue.shift()!;
+      await this.processNode(current.node, variables, current.parent);
+
+      if (current.node.content) {
+        for (const childNode of current.node.content) {
+          queue.push({ node: childNode, parent: current.node });
         }
       }
-      node.content = processedContent;
     }
+  }
 
+  private async processNode(node: TipTapNode, variables: Record<string, unknown>, parent?: TipTapNode): Promise<void> {
     if (this.hasShow(node)) {
-      await this.hideShowIfNeeded(variables, node, parentNode);
-    } else if (this.hasEach(node)) {
-      const newContent = this.expendedForEach(node);
-      node.content = newContent;
-      if (parentNode && parentNode.content) {
-        this.insertArrayAt(parentNode.content, parentNode.content.indexOf(node), newContent);
-        parentNode.content.splice(parentNode.content.indexOf(node), 1);
-      }
+      await this.handleShowNode(node, variables, parent);
     }
 
-    return node;
-  }
-  private insertArrayAt(array: any[], index: number, newArray: any[]) {
-    if (index < 0 || index > array.length) {
-      throw new Error('Index out of bounds');
+    if (this.hasEach(node)) {
+      await this.handleEachNode(node, variables, parent);
     }
-    array.splice(index, 0, ...newArray);
   }
 
-  private hasEach(node: TipTapNode): node is TipTapNode & { attrs: { each: unknown } } {
-    return !!(node.attrs && 'each' in node.attrs);
+  private async handleShowNode(
+    node: TipTapNode & { attrs: { showIfKey: unknown } },
+    variables: Record<string, unknown>,
+    parent?: TipTapNode
+  ): Promise<void> {
+    const shouldShow = await this.evaluateShowCondition(variables, node);
+    if (!shouldShow && parent?.content) {
+      parent.content = parent.content.filter((pNode) => pNode !== node);
+
+      return;
+    }
+    if (node.attrs) {
+      delete node.attrs[MailyAttrsEnum.SHOW_IF_KEY];
+    }
   }
 
-  private hasShow(node: TipTapNode): node is TipTapNode & { attrs: { [MailyAttrsEnum.SHOW_IF_KEY]: string } } {
-    return node.attrs?.[MailyAttrsEnum.SHOW_IF_KEY] !== undefined;
+  private async handleEachNode(
+    node: TipTapNode & { attrs: { each: unknown } },
```

Comment:
Ditto about enhancing the TipTapNode shared type

Reply:
I believe in this context we are talking about enhancing Maily JSONContent (TipTap based) opinionated type